core/controllers/async_download.py

```diff
@@ -91,69 +91,74 @@ def async_download(
-
+    unique_id = str(uuid.uuid4())
+    file_name = f"fund-monitoring-data-{current_datetime}-{unique_id}.{file_format}"
     find_service_base_url = Config.FIND_SERVICE_BASE_URL
+    download_url = f"{find_service_base_url}/retrieve-download/{file_name}"
+    find_service_download_url: str = find_service_base_url + "/download"
+
     if not find_service_base_url:
         raise ValueError("FIND_SERVICE_BASE_URL is not set.")
 
-    find_service_download_url: str = find_service_base_url + "/download"
-    upload_find_file(file=file_obj, bucket=bucket, object_name=file_name)
-    presigned_url = generate_find_presigned_url(bucket=bucket, object_name=file_name, expiry=86400)
+    try:
+        upload_file(file=file_obj, bucket=bucket, object_name=file_name)
+    except KeyError:
+        current_app.logger.error("Failed to upload file to S3: {e}")
+        return
 
-    # email the URL to the user
     send_email_for_find_download(
         email_address=email_address,
-        download_url=presigned_url,
+        download_url=download_url,
         find_service_url=find_service_download_url,
-        file_format=file_format_str,
-        file_size_str=file_size_str,
     )
 
 
-def upload_find_file(file: IO | FileStorage, bucket: str, object_name: str):
-    """Upload a file to an S3 bucket
-    :param file: File to upload,
-    :param bucket: Bucket to upload to,
-    :param object_name: S3 object name (filename),
-    """
-
-    _S3_CLIENT.put_object(Bucket=bucket, Key=object_name, Body=file)
-
+def get_find_download_file_metadata(filename):
+    """Retrieve metadata about a file in S3."""
 
-def generate_find_presigned_url(bucket: str, object_name: str, expiry: int) -> str:
-    """Generate a presigned URL for an S3 object.
-    :param bucket: S3 bucket name,
-    :param object_name: S3 object name,
-    :param expiry: URL expiry time in seconds,
+    try:
+        response = _S3_CLIENT.head_object(Bucket=Config.AWS_S3_BUCKET_FIND_DATA_FILES, Key=filename)
+        metadata = {
+            "content_length": response["ContentLength"],
+            "content_type": response["ContentType"],
+            "last_modified": response["LastModified"].isoformat(),
+        }
+        return jsonify(metadata)
+    except ClientError as error:
+        if error.response["Error"]["Code"] == "404":
+            return abort(404, f"Could not find file: {filename} on S3.")
+        raise error
+
+
+def get_presigned_url(filename: str):
+    """retrieve a file a from S3 bucket if exist"""
 
-    :return: presigned URL,
-    """
+    try:
+        # Check if the object (file) exists in S3
+        _S3_CLIENT.head_object(Bucket=Config.AWS_S3_BUCKET_FIND_DATA_FILES, Key=filename)
+    except ClientError as error:
+        if error.response["Error"]["Code"] == "404":
+            return abort(404, f"Could not find file: {filename} on S3.")
+        else:
+            return abort(500, f"Error checking object: {filename} existence in S3")
```

Comment:
We should ideally have a consistent data type as the return value. We generally don't just return raw strings in JSON responses - can we at least wrap these in a dictionary, like the "happy path" response?

Something like: `return jsonify({"error": {"code": "file-not-found", "message": f"Could not find file {filename} in S3."}}), 404`